src/precompile.c

```diff
@@ -0,0 +1,363 @@
+            return 1; // why does this method exist?
+    }
+
+    if (count_unions == 0)
+        return _compile_all_tvar_union(sig);
+
+    int *idx = (int*)alloca(sizeof(int) * count_unions);
+    for (i = 0; i < count_unions; i++) {
+        idx[i] = 0;
+    }
+
+    JL_GC_PUSH2(&p, &methsig);
+    int idx_ctr = 0, incr = 0;
+    while (!incr) {
+        jl_svec_t *p = jl_alloc_svec_uninit(l);
+        for (i = 0, idx_ctr = 0, incr = 1; i < l; i++) {
+            jl_value_t *ty = jl_svecref(sigbody->parameters, i);
+            if (jl_is_uniontype(ty)) {
+                size_t l = jl_count_union_components(ty);
+                size_t j = idx[idx_ctr];
+                jl_svecset(p, i, jl_nth_union_component(ty, j));
+                ++j;
+                if (incr) {
+                    if (j == l) {
+                        idx[idx_ctr] = 0;
+                    }
+                    else {
+                        idx[idx_ctr] = j;
+                        incr = 0;
+                    }
+                }
+                ++idx_ctr;
+            }
+            else {
+                jl_svecset(p, i, ty);
+            }
+        }
+        methsig = (jl_value_t*)jl_apply_tuple_type(p);
+        methsig = jl_rewrap_unionall(methsig, sig);
+        if (!_compile_all_tvar_union(methsig))
+            complete = 0;
+    }
+
+    JL_GC_POP();
+    return complete;
+}
+
+static void _compile_all_deq(jl_array_t *found)
+{
+    int found_i, found_l = jl_array_len(found);
+    jl_printf(JL_STDERR, "found %d uncompiled methods for compile-all\n", (int)found_l);
+    jl_method_instance_t *linfo = NULL;
+    jl_value_t *src = NULL;
+    JL_GC_PUSH2(&linfo, &src);
+    for (found_i = 0; found_i < found_l; found_i++) {
+        if (found_i % (1 + found_l / 300) == 0 || found_i == found_l - 1) // show 300 progress steps, to show progress without overwhelming log files
+            jl_printf(JL_STDERR, " %d / %d\r", found_i + 1, found_l);
+        jl_typemap_entry_t *ml = (jl_typemap_entry_t*)jl_array_ptr_ref(found, found_i);
+        jl_method_t *m = ml->func.method;
+        if (m->isstaged)  // TODO: generic implementations of generated functions
+            continue;
+        linfo = m->unspecialized;
+        if (!linfo) {
+            linfo = jl_get_specialized(m, (jl_value_t*)m->sig, jl_emptysvec);
+            m->unspecialized = linfo;
+            jl_gc_wb(m, linfo);
+        }
+
+        if (linfo->jlcall_api == 2)
+            continue;
+        src = m->source;
+        // TODO: the `unspecialized` field is not yet world-aware, so we can't store
+        // an inference result there.
+        //src = jl_type_infer(&linfo, jl_world_counter, 1);
+        //m->unspecialized = linfo;
+        //jl_gc_wb(m, linfo);
+        //if (linfo->jlcall_api == 2)
+        //    continue;
+
+        _compile_all_union((jl_value_t*)ml->sig);
+        // if _compile_all_union returns 1, all possible signatures have been covered, so we don't
```

Comment:
I don't quite get it. I thought the idea of the return value of `_compile_all_union` was that we didn't need to infer or compile an unspecialized version, since the signature is equivalent to a set of concrete signatures.